GH-39001: [Java] Modularize remaining modules

jduo opened the pull request:
### Rationale for this change
Modularize remaining modules outside of memory modules, vector, and format.

### What changes are included in this PR?

### Are these changes tested?
Yes, existing unit tests now run with modules when using JDK9+.

### Are there any user-facing changes?
Yes. There are new command-line options that may be necessary. The way of specifying the output directory for
JNI native library builds differs. The flight-grpc module has been eliminated since it is now built into flight-core.
Documentation has been updated for these changes.

**This PR includes breaking changes to public APIs.**
There are a number of package structure changes and some modules now need additional command-line arguments.

<!-- **This PR contains a "Critical Fix".** -->
* Closes: #39001

jduo commented:
dataset is having issues due to the directory holding native libs in JARs (the arch eg x86_64) is being treated as a package name. flight-core and other Flight modules requires additional command-line arguments.

Netty is omitted due to requiring user-unfriendly command-line arguments.

jduo commented:
Doc changes are required as different modules require more command line arguments. I'll update the docs.

assignUser commented:
FYI for testing runing the java-jars job might be useful. You can start it by commenting `@github-action crossbow submit java-jars`

jduo commented:
> FYI for testing runing the java-jars job might be useful. You can start it by commenting `@github-action crossbow submit java-jars`

Thanks, I'll run that now.

jduo commented:
@github-action crossbow *submit* *java-jars*

jduo commented:
@github-action crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
This is failing the integration test currently because it cannot find the native library:
https://github.com/ursacomputing/crossbow/actions/runs/7495454991/job/20405564594#step:6:9647

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@lidavidm @assignUser @raulcd @danepitkin The checks that have failed now seem to be flakey.

The java-jars job is failing when verifying JARs with a strange "not a git repository" error.

raulcd commented:
> The java-jars job is failing when verifying uploaded JARs with a strange "not a git repository" error.

The git repository message is not making the job fail, the current error is that it seems it can't download some of the expected generated packages:
```
                           flight-grpc-15.0.0-SNAPSHOT-cyclonedx.json [PENDING]
                            flight-grpc-15.0.0-SNAPSHOT-cyclonedx.xml [PENDING]
                              flight-grpc-15.0.0-SNAPSHOT-javadoc.jar [PENDING]
                              flight-grpc-15.0.0-SNAPSHOT-sources.jar [PENDING]
                                flight-grpc-15.0.0-SNAPSHOT-tests.jar [PENDING]
                                      flight-grpc-15.0.0-SNAPSHOT.jar [PENDING]
                                      flight-grpc-15.0.0-SNAPSHOT.pom [PENDING]
```
It seems the PR is removing those:
```
      <dependency>
        <groupId>org.apache.arrow</groupId>
        <artifactId>flight-grpc</artifactId>
        <version>${project.version}</version>
      </dependency>
```
We have to remove them from the job here: https://github.com/apache/arrow/blob/main/dev/tasks/tasks.yml#L813-L819
if that's the expectation and those should not be generated anymore.

kou commented on ci/scripts/java_jni_build.sh:
```diff
@@ -56,7 +68,7 @@ cmake \
   -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} \
   -DCMAKE_PREFIX_PATH=${arrow_install_dir} \
-  -DCMAKE_INSTALL_LIBDIR=lib \
+  -DCMAKE_ARCH_DIR=${normalized_arch} \
```
Please don't use `CMAKE_` prefix for our variable. Please use `ARROW_JAVA_JNI_` prefix like other our variables such as `ARROW_JAVA_JNI_ENABLE_DEFAULT` instead.

BTW, why do we need to install to sub directory such as `lib/arrow_cdata_jni/${arch}/libarrow_cdata_jni.so` instead of `lib/${arch}/libarrow_cdata_jni.so`? Nameing conflict? We don't have distinct base name such `libarrow_cdata_jni` and `libarrow_dataset_jni`. So I think that no naming conflict is happen.

jduo replied:
Sure, I'll fix the name of the ARCH_DIR variable.

Yes the reason is a naming conflict. It's only visible when using JPMS modules.
It's because we see multiple JPMS modules (where a JPMS module is a JAR with a module-info.class file) that contain the same subfolder at the root level (the ${arch} folder). JPMS sees this as a package collision even though the directory only contains binaries and throws startup exceptions in this scenario.

jduo commented:
> > The java-jars job is failing when verifying uploaded JARs with a strange "not a git repository" error.
> 
> The git repository message is not making the job fail, the current error is that it seems it can't download some of the expected generated packages:
> 
> ```
>                            flight-grpc-15.0.0-SNAPSHOT-cyclonedx.json [PENDING]
>                             flight-grpc-15.0.0-SNAPSHOT-cyclonedx.xml [PENDING]
>                               flight-grpc-15.0.0-SNAPSHOT-javadoc.jar [PENDING]
>                               flight-grpc-15.0.0-SNAPSHOT-sources.jar [PENDING]
>                                 flight-grpc-15.0.0-SNAPSHOT-tests.jar [PENDING]
>                                       flight-grpc-15.0.0-SNAPSHOT.jar [PENDING]
>                                       flight-grpc-15.0.0-SNAPSHOT.pom [PENDING]
> ```
> 
> It seems the PR is removing those:
> 
> ```
>       <dependency>
>         <groupId>org.apache.arrow</groupId>
>         <artifactId>flight-grpc</artifactId>
>         <version>${project.version}</version>
>       </dependency>
> ```
> 
> We have to remove them from the job here: https://github.com/apache/arrow/blob/main/dev/tasks/tasks.yml#L813-L819 if that's the expectation and those should not be generated anymore.

Thanks. Yes flight-grpc got merged into the flight-core module.

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@raulcd , this error showed up relating to Arrow BOM:
https://github.com/ursacomputing/crossbow/actions/runs/7531736017/job/20501955371#step:6:3047

I rebased as well, so could it be related to #39580 ?

jduo commented:
> @raulcd , this error showed up relating to Arrow BOM: https://github.com/ursacomputing/crossbow/actions/runs/7531736017/job/20501955371#step:6:3047
> 
> I rebased as well, so could it be related to #39580 ?

This error seemed to have went away by itself (from the same job even). So now java-jars is green.

lidavidm commented:
Are we ok with the CMake changes here? CC @vibhatha @davisusanibar too

vibhatha commented:
@lidavidm I am still reviewing, will confirm.

kou commented:
(I'll re-review the CMake part. Please wait for a few days...)

vibhatha commented:
CMake changes looks fine to me. @kou appreciate your feedback in a few days.

kou commented on ci/scripts/java_jni_build.sh:
```diff
@@ -24,6 +24,18 @@ arrow_install_dir=${2}
 build_dir=${3}/java_jni
 # The directory where the final binaries will be stored when scripts finish
 dist_dir=${4}
+normalized_arch=$(arch)
+case ${normalized_arch} in
+  aarch64)
+    normalized_arch=aarch_64
+    ;;
+  i386)
+    normalized_arch=x86_64
+    ;;
+  arm64)
+    normalized_arch=aarch_64
+    ;;
+  esac
```
Can we do this in `java/CMakeLists.txt` instead of specifying this by a user?
For example, can we use `CMAKE_HOST_SYSTEM_PROCESSOR` https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_PROCESSOR.html for this?

jduo replied:
I used CMAKE_HOST_SYSTEM_PROCESSOR now.

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented on java/gandiva/CMakeLists.txt:
```diff
@@ -84,6 +84,11 @@ if(CXX_LINKER_SUPPORTS_VERSION_SCRIPT)
+set(CMAKE_INSTALL_LIBDIR
+    "${CMAKE_INSTALL_PREFIX}/lib/gandiva_jni/${ARROW_JAVA_JNI_ARCH_DIR}")
+set(CMAKE_INSTALL_BINDIR
+    "${CMAKE_INSTALL_PREFIX}/bin/gandiva_jni/${ARROW_JAVA_JNI_ARCH_DIR}")
```
Done.

jduo commented:
@kou I've addressed your comments. I haven't been able to run the java-jars job though because there's a failure pulling S3 when compiling on Mac.

kou commented on ci/scripts/java_jni_build.sh:
```diff
@@ -56,8 +55,7 @@ cmake \
   -DCMAKE_PREFIX_PATH=${arrow_install_dir} \
-  -DCMAKE_INSTALL_LIBDIR=lib \
-  -DCMAKE_INSTALL_PREFIX=${prefix_dir} \
+\  -DCMAKE_INSTALL_PREFIX=${prefix_dir} \
```
```suggestion
  -DCMAKE_INSTALL_PREFIX=${prefix_dir} \
```

kou commented on ci/scripts/java_jni_macos_build.sh:
```diff
@@ -82,7 +82,7 @@ cmake \
   -DARROW_USE_CCACHE=${ARROW_USE_CCACHE} \
   -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} \
-  -DCMAKE_INSTALL_LIBDIR=lib \
+  -DARROW_JAVA_JNI_ARCH_DIR=${normalized_arch} \
```
Can we remove this and `normalized_arch` from this script?

kou commented on ci/scripts/java_jni_manylinux_build.sh:
```diff
@@ -91,7 +91,7 @@ cmake \
   -DARROW_USE_CCACHE=${ARROW_USE_CCACHE} \
   -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} \
-  -DCMAKE_INSTALL_LIBDIR=lib \
+  -DARROW_JAVA_JNI_ARCH_DIR=${normalized_arch} \
```
Can we remove this and `normalized_arch` from this script?

kou commented on ci/scripts/java_jni_windows_build.sh:
```diff
@@ -72,7 +72,7 @@ cmake \
   -DARROW_WITH_ZSTD=ON \
   -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} \
-  -DCMAKE_INSTALL_LIBDIR=lib \
+  -DARROW_JAVA_JNI_ARCH_DIR=x86_64 \
```
Can we remove this and `normalized_arch` from this script?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -166,12 +165,11 @@ CMake
           -DBUILD_TESTING=OFF \
           -DCMAKE_BUILD_TYPE=Release \
-          -DCMAKE_INSTALL_LIBDIR=lib/<your system's architecture> \
+          -DARROW_JAVA_JNI_ARCH_DIR=<your system's architecture> \
```
Can we remove this?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -186,11 +184,11 @@ CMake
           -DBUILD_TESTING=OFF ^
           -DCMAKE_BUILD_TYPE=Release ^
-          -DCMAKE_INSTALL_LIBDIR=lib/x86_64 ^
+          -DARROW_JAVA_JNI_ARCH_DIR=x86_64 ^
```
Can we remove this?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -222,7 +220,7 @@ CMake
           -DARROW_USE_CCACHE=ON \
           -DCMAKE_BUILD_TYPE=Release \
-          -DCMAKE_INSTALL_LIBDIR=lib/<your system's architecture> \
+          -DARROW_JAVA_JNI_ARCH_DIR=<your system's architecture> \
```
Can we remove this?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -233,16 +231,16 @@ CMake
           -DBUILD_TESTING=OFF \
           -DCMAKE_BUILD_TYPE=Release \
-          -DCMAKE_INSTALL_LIBDIR=lib/<your system's architecture> \
+          -DARROW_JAVA_JNI_ARCH_DIR=<your system's architecture> \
```
Can we remove this?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -271,7 +269,7 @@ CMake
           -DARROW_WITH_ZSTD=ON ^
           -DCMAKE_BUILD_TYPE=Release ^
-          -DCMAKE_INSTALL_LIBDIR=lib/x86_64 ^
+          -DARROW_JAVA_JNI_ARCH_DIR=x86_64 ^
```
Can we remove this?

kou commented on docs/source/developers/java/building.rst:
```diff
@@ -288,13 +286,13 @@ CMake
           -DBUILD_TESTING=OFF ^
           -DCMAKE_BUILD_TYPE=Release ^
-          -DCMAKE_INSTALL_LIBDIR=lib/x86_64 ^
+          -DARROW_JAVA_JNI_ARCH_DIR=x86_64 ^
```
Can we remove this?

kou commented on java/CMakeLists.txt:
```diff
@@ -74,6 +74,20 @@ if(BUILD_TESTING)
                                                   GTest::gtest_main)
 endif()
 
+# If the user hasn't set ARROW_JAVA_JNI_ARCH_DIR, derive the normalized
```
In general, should the user specify this?
Can we use auto detection as the recommended approach? If the auto detection failed, the user can specify this explicitly as a workaround.

jduo replied:
I'll change the phrasing of this comment to suggest that the user shouldn't normally need to set this.

kou commented on java/CMakeLists.txt:
```diff
@@ -74,6 +74,20 @@ if(BUILD_TESTING)
+# If the user hasn't set ARROW_JAVA_JNI_ARCH_DIR, derive the normalized
+# operating system from the host processor.
+if("${ARROW_JAVA_JNI_ARCH_DIR}" STREQUAL "")
+  if("${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "aarch64")
```
Sorry.
`CMAKE_SYSTEM_PROCESSOR` https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PROCESSOR.html is suitable for this.
It's the processor for build target.

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@github-actions crossbow submit *java-jars*

jduo commented:
@kou , I've switched to using CMAKE_SYSTEM_PROCESSOR and changed CI scripts and POM files to let auto-detection figure ARROW_JAVA_JNI_ARCH_DIR. I think this addresses everything.

kou reviewed (approved):
+1 for the CMake part

raulcd commented:
@github-actions crossbow submit verify-rc-source-integration-linux-*

raulcd commented:
I am verifying if this caused: https://github.com/apache/arrow/issues/41201
